### ISSUE TITLE
[non-nullable-type-assertion-style] false positive when using type assertion to specify type

### ISSUE THREAD

**Issue by yangmingshan** [typescript-eslint/typescript-eslint#2887]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/non-nullable-type-assertion-style": "error"
  }
}
```

```TS
function inject<T>(key: string): T | undefined {
  // ...
}

// Of course, I can write it to `const str = inject<string>('myString')!`, but still seems a bug.
// Autofix will fix it to `const str = inject('myString')!`, and might break type check.
const str = inject('myString') as string
```

tsconfig

```json
{
  "compilerOptions": {
    "target": "esnext",
    "lib": ["esnext"],
    "skipLibCheck": true,
    "esModuleInterop": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true
  }
}
```

**Expected Result**

No error.

**Actual Result**

Use a ! assertion to more succintly remove null and undefined from the type.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.10.0` |
| `@typescript-eslint/parser`        | `4.10.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `15.4.0` |

**Comment by bradzacher** [comment]
This is a known problem with functions that have generics that are only used in the return position.

There's nothing we can do about it.

In a nutshell.

```
declare function foo<T>(): T;

const a = foo(); // T is unknown
const b = foo<string>(); // T is string
const c: string = foo(); // T is string
const d = foo() as string; // T is string
```

TypeScript essentially "backfills" the generic in the return position based on the type assertion applied to it if there is no explicit generic provided.

So when we ask TS what the return type of `foo()` is when it is followed by `as string`, it tells us `string`, even though it *should* be `unknown`.

This causes a the bug. There's the same problem with the no-unnecessary-type-assertion rule.

I've spoken to TS about this and their response is that it is what it is and it is just another reason not to use the unsafe anti-pattern of a generic that's only used in the return position.

**Comment by yangmingshan** [comment]
@bradzacher So the autofix is not always reliable, are you consider change this rule to `suggestion`?
